metrics/config_test.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2018 The Knative Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package metrics
+
+import (
+	"testing"
+	"time"
+
+	. "github.com/knative/pkg/logging/testing"
+)
+
+const (
+	testProj      = "test-project"
+	testDomain    = "test.domain"
+	testComponent = "testComponent"
+)
+
+func TestNewStackdriverExporter(t *testing.T) {
+	// The stackdriver project ID is required for stackdriver exporter.
+	err := newMetricsExporter(metricsConfig{
+		domain:               testDomain,
+		component:            testComponent,
+		backendDestination:   Stackdriver,
+		stackdriverProjectID: testProj}, TestLogger(t))
+	if err != nil {
+		t.Error(err)
+	}
+	expectNoPromSrv(t)
+}
+
+func TestNewPrometheusExporter(t *testing.T) {
+	// The stackdriver project ID is not required for prometheus exporter.
+	err := newMetricsExporter(metricsConfig{
+		domain:               testDomain,
+		component:            testComponent,
+		backendDestination:   Prometheus,
+		stackdriverProjectID: ""}, TestLogger(t))
+	if err != nil {
+		t.Error(err)
+	}
+	expectPromSrv(t)
+}
+
+func TestInterlevedExporters(t *testing.T) {
+	// First create a stackdriver exporter
+	err := newMetricsExporter(metricsConfig{
+		domain:               testDomain,
+		component:            testComponent,
+		backendDestination:   Stackdriver,
+		stackdriverProjectID: testProj}, TestLogger(t))
+	if err != nil {
+		t.Error(err)
+	}
+	err = newMetricsExporter(metricsConfig{
+		domain:               testDomain,
+		component:            testComponent,
+		backendDestination:   Prometheus,
+		stackdriverProjectID: "",
+	}, TestLogger(t))
+	if err != nil {
+		t.Error(err)
+	}
+	expectPromSrv(t)
+	// Finally switch to stackdriver exporter
+	err = newMetricsExporter(metricsConfig{
+		domain:               testDomain,
+		component:            testComponent,
+		backendDestination:   Stackdriver,
+		stackdriverProjectID: testProj}, TestLogger(t))
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetMetricsConfig(t *testing.T) {
+	_, err := getMetricsConfig(map[string]string{
```

Comment:
Config tests should be mostly table tests. See https://github.com/knative/serving/blob/master/pkg/autoscaler/config_test.go as an example.